lib/Dialect/TritonGPU/Transforms/Pipeliner/PipeliningUtility.cpp

```diff
@@ -643,7 +643,24 @@ ttg::SharedEncodingTrait mlir::triton::getSharedEncoding(Operation *op) {
   if (localAllocEnc)
     return localAllocEnc;
 
-  // Use generic layout. This won't be optimal for 2D tensors.
+  // Use generic layout but ensure CTALayout matches the source pointer's layout
+  // when available, so that src/dst CTALayouts are identical for cp.async.
+  // This helps make the convert layout trivial over the `block` dimension.
+  if (auto load = dyn_cast<tt::LoadOp>(op)) {
+    Value srcPtr = load.getPtr();
+    Type srcPtrTy = srcPtr.getType();
+    if (auto ptrTy = dyn_cast<triton::PointerType>(srcPtrTy)) {
+      Type pointee = ptrTy.getPointeeType();
+      if (auto rankedTensor = dyn_cast<RankedTensorType>(pointee)) {
+        if (Attribute srcEncAttr = rankedTensor.getEncoding()) {
+          if (auto srcCta = ttg::getCTALayout(srcEncAttr)) {
```

Comment:
Can any of these ifs ever fail?

Reply:
fixed.

Reply:
how has this been fixed? My point is that all these ifs look like they always be true no? What are counterexamples for a LoadOp not having a ptr type or a ptr not to a ranked tensor?

Reply:
Sorry, I missed your point. Now I see! I’ll double-check for counterexamples and handle corner cases found.

Reply:
We don't have tensors of pointers in Triton?
Also tensors do have an encoding at this point.

I am not sure we would allow to have just one pointer in cp.async. And if we do, we probably should error out in the checker, as this can always be represented as a pointer to a 1 element tensor.